src/neural/cuda/layers.cc

```diff
@@ -1225,6 +1187,431 @@ ResidualBlock<DataType>::~ResidualBlock() {
+  ReportCUDAErrors(
+      cudaMemcpy(scratch, &cpu_src[0], cpu_src.size() * sizeof(float), cudaMemcpyHostToDevice));
+  copyTypeConverted((DataType*)(*gpu_dest), (float*)scratch,
+                    (int)cpu_src.size(), 0);
+}
+
+template <typename DataType>
+AttentionPolicyHead<DataType>::AttentionPolicyHead(BaseLayer<DataType>* ip,
+                                                   const LegacyWeights& weights,
+                                                   void* scratch)
+    : BaseLayer<DataType>(64 * 64 + 24 * 8, 1, 1, ip) {
+  embedding_op_size_ = weights.ip_pol_b.size();
+  wq_op_size_ = weights.ip2_pol_b.size();
+  wk_op_size_ = weights.ip3_pol_b.size();
+
+  encoder_heads_ = weights.encoder_head_count;
+  policy_d_model_ = wq_op_size_;
+
+  allocAndUpload<DataType>(&ip_pol_w_, weights.ip_pol_w, scratch);
+  allocAndUpload<DataType>(&ip_pol_b_, weights.ip_pol_b, scratch);
+
+  allocAndUpload<DataType>(&ip2_pol_w_, weights.ip2_pol_w, scratch);
+  allocAndUpload<DataType>(&ip2_pol_b_, weights.ip2_pol_b, scratch);
+
+  allocAndUpload<DataType>(&ip3_pol_w_, weights.ip3_pol_w, scratch);
+  allocAndUpload<DataType>(&ip3_pol_b_, weights.ip3_pol_b, scratch);
+
+  allocAndUpload<DataType>(&ip4_pol_w_, weights.ip4_pol_w, scratch);
+
+  for (const auto& enc : weights.encoder) {
+    EncoderWeights* pW = new EncoderWeights(enc, scratch);
+    encoder_weights_.emplace_back(pW);
+  }
+}
+
+template <typename DataType>
+AttentionPolicyHead<DataType>::EncoderWeights::EncoderWeights(
+    const LegacyWeights::EncoderLayer& cpu_weights, void* scratch) {
+  mha_q_size_ = cpu_weights.mha.q_b.size();
+  mha_k_size_ = cpu_weights.mha.k_b.size();
+  mha_v_size_ = cpu_weights.mha.v_b.size();
+  mha_dense_size_ = cpu_weights.mha.dense_b.size();
+  ffn_dense1_size_ = cpu_weights.ffn.dense1_b.size();
+  ffn_dense2_size_ = cpu_weights.ffn.dense2_b.size();
+
+  // debug!
+  printf("\nsize of weight mha.q_b/w: %d, %d\n",
+         (int)cpu_weights.mha.q_b.size(), (int)cpu_weights.mha.q_w.size());
+  printf("\nsize of weight mha.k_b/w: %d, %d\n",
+         (int)cpu_weights.mha.k_b.size(), (int)cpu_weights.mha.k_w.size());
+  printf("\nsize of weight mha.v_b/w: %d, %d\n",
+         (int)cpu_weights.mha.v_b.size(), (int)cpu_weights.mha.v_w.size());
+  printf("\nsize of weight mha.dense_b/w: %d, %d\n",
+         (int)cpu_weights.mha.dense_b.size(),
+         (int)cpu_weights.mha.dense_w.size());
+  printf("\nsize of ln1 betas/gammas: %d, %d\n",
+         (int)cpu_weights.ln1_betas.size(), (int)cpu_weights.ln1_gammas.size());
+  printf("\nsize of ln2 betas/gammas: %d, %d\n",
+         (int)cpu_weights.ln2_betas.size(), (int)cpu_weights.ln2_gammas.size());
+  printf("\nsize of weight ffn.dense1_b/w: %d, %d\n",
+         (int)cpu_weights.ffn.dense1_b.size(),
+         (int)cpu_weights.ffn.dense1_w.size());
+  printf("\nsize of weight ffn.dense2_b/w: %d, %d\n",
+         (int)cpu_weights.ffn.dense2_b.size(),
+         (int)cpu_weights.ffn.dense2_w.size());
```

Comment:
these debugs are still here (third attempt to highlight the code correctly)

Reply:
thanks. Removed them.